js/modules/loki_snode_api.js

```diff
@@ -35,77 +36,122 @@ class LokiSnodeAPI {
-      }));
-    } catch (e) {
-      log.warn('initialiseRandomPool error', e.code, e.message);
-      if (consecutiveErrors < 3) {
-        // retry after a possible delay
-        setTimeout(() => {
+    // if currently not in progress
+    if (this.initialiseRandomPoolPromise === false) {
+      // FIXME: add timeout
+      // set lock
+      this.initialiseRandomPoolPromise = new Promise(async resolve => {
+        const params = {
+          limit: 1024,
+          active_only: true,
+          fields: {
+            public_ip: true,
+            storage_port: true,
+            pubkey_x25519: true,
+            pubkey_ed25519: true,
+          },
+        };
+        const seedNode = seedNodes.splice(
+          Math.floor(Math.random() * seedNodes.length),
+          1
+        )[0];
+        let snodes = [];
+        try {
+          log.info('loki_snodes: Refreshing random snode pool');
+          const response = await lokiRpc(
+            `http://${seedNode.ip}`,
+            seedNode.port,
+            'get_n_service_nodes',
+            params,
+            {}, // Options
+            '/json_rpc' // Seed request endpoint
+          );
+          // Filter 0.0.0.0 nodes which haven't submitted uptime proofs
+          snodes = response.result.service_node_states.filter(
+            snode => snode.public_ip !== '0.0.0.0'
+          );
+          this.randomSnodePool = snodes.map(snode => ({
+            ip: snode.public_ip,
+            port: snode.storage_port,
+            pubkey_x25519: snode.pubkey_x25519,
+            pubkey_ed25519: snode.pubkey_ed25519,
+          }));
           log.info(
-            'Retrying initialising random snode pool, try #',
-            consecutiveErrors
+            'loki_snodes: Refreshed random snode pool with',
+            this.randomSnodePool.length,
+            'snodes'
           );
-          this.initialiseRandomPool(seedNodes, consecutiveErrors + 1);
-        }, consecutiveErrors * consecutiveErrors * 5000);
-      } else {
-        log.error('Giving up trying to contact seed node');
-        if (snodes.length === 0) {
-          throw new window.textsecure.SeedNodeError(
-            'Failed to contact seed node'
+        } catch (e) {
+          log.warn(
+            'loki_snodes: initialiseRandomPool error',
+            e.code,
+            e.message
           );
+          if (consecutiveErrors < 3) {
+            // retry after a possible delay
+            setTimeout(() => {
+              log.info(
+                'loki_snodes: Retrying initialising random snode pool, try #',
+                consecutiveErrors
+              );
+              this.initialiseRandomPool(seedNodes, consecutiveErrors + 1);
+            }, consecutiveErrors * consecutiveErrors * 5000);
+          } else {
+            log.error('loki_snodes: Giving up trying to contact seed node');
+            if (snodes.length === 0) {
+              throw new window.textsecure.SeedNodeError(
```

Comment:
Don't we need to reset `this.initialiseRandomPoolPromise` here?

Reply:
Yes, it stops 6 requests asking for 1024 nodes to the seed node on startup in my client.